qcodes/tests/test_delegate_parameter.py

```diff
@@ -294,3 +296,115 @@ def test_setting_initial_cache_delegate_parameter():
                           initial_cache_value=value)
     assert p.cache.get(get_if_invalid=False) == value
     assert d.cache.get(get_if_invalid=False) == value
+
+
+def test_delegate_parameter_with_none_source_works_as_expected():
+    delegate_param = DelegateParameter(name='delegate', source=None,
+                                       scale=2, offset=1)
+    _assert_none_source_is_correct(delegate_param)
+
+
+@given(hst.floats(allow_nan=False, allow_infinity=False),
+       hst.floats(allow_nan=False, allow_infinity=False).filter(lambda x: x != 0),
+       hst.floats(allow_nan=False, allow_infinity=False))
+def test_delegate_parameter_with_changed_source_snapshot_matches_value(value,
+                                                                       scale,
+                                                                       offset):
+    delegate_param = DelegateParameter(name="delegate",
+                                       source=None,
+                                       scale=scale,
+                                       offset=offset)
+    source_parameter = Parameter(name="source",
+                                 get_cmd=None,
+                                 set_cmd=None,
+                                 initial_value=value)
+    _assert_none_source_is_correct(delegate_param)
+    delegate_param.source = source_parameter
+    calc_value = (value - offset) / scale
+    assert delegate_param.cache.get(get_if_invalid=False) == calc_value
+    assert delegate_param.source.cache.get(get_if_invalid=False) == value
+    snapshot = delegate_param.snapshot()
+    # disregard timestamp that might be slightly different
+    snapshot["source_parameter"].pop("ts")
+    source_snapshot = source_parameter.snapshot()
+    source_snapshot.pop("ts")
+    assert snapshot["source_parameter"] == source_snapshot
+    assert snapshot["value"] == calc_value
+    assert delegate_param.get() == calc_value
+    # now remove the source again
+    delegate_param.source = None
+    _assert_none_source_is_correct(delegate_param)
+    _assert_delegate_cache_none_source(delegate_param)
+
+
+def _assert_none_source_is_correct(delegate_param):
+    with pytest.raises(TypeError):
+        delegate_param.get()
+    with pytest.raises(TypeError):
+        delegate_param.set(1)
+    snapshot = delegate_param.snapshot()
+    assert snapshot["source_parameter"] is None
+    assert "value" not in snapshot.keys()
+    snapshot.pop("ts")
+    updated_snapshot = delegate_param.snapshot(update=True)
+    updated_snapshot.pop("ts")
+    assert snapshot == updated_snapshot
+
+
+def _assert_delegate_cache_none_source(delegate_param):
+    with pytest.raises(TypeError):
+        delegate_param.cache.set(1)
+    with pytest.raises(TypeError):
+        delegate_param.cache.get()
+    with pytest.raises(TypeError):
+        delegate_param.cache.raw_value()
```

Comment:
it's a property, right? not a method
```suggestion
        delegate_param.cache.raw_value
```